packages/editor/src/components/provider/index.js

```diff
@@ -37,36 +37,59 @@ import ConvertToGroupButtons from '../convert-to-group-buttons';
  * It seems like there is no suitable package to import this from. Ideally it would be either part of core-data.
  * Until we refactor it, just copying the code is the simplest solution.
  *
- * @param {Object} search
- * @param {number} perPage
+ * @param {string} search
+ * @param {Object} arguments
+ * @param {number} [arguments.perPage=20]
+ * @param {string} [arguments.type='post']
+ * @param {string} [arguments.subtype]
  * @return {Promise<Object[]>} List of suggestions
  */
-const fetchLinkSuggestions = async ( search, { perPage = 20 } = {} ) => {
-	const posts = apiFetch( {
-		path: addQueryArgs( '/wp/v2/search', {
-			search,
-			per_page: perPage,
-			type: 'post',
-		} ),
-	} );
+const fetchLinkSuggestions = async (
+	search,
+	{ perPage = 20, type, subtype } = {}
+) => {
+	const queries = [];
 
-	const terms = apiFetch( {
-		path: addQueryArgs( '/wp/v2/search', {
-			search,
-			per_page: perPage,
-			type: 'term',
-		} ),
-	} );
+	if ( ! type || type === 'post' ) {
+		queries.push(
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/search', {
+					search,
+					per_page: perPage,
+					type: 'post',
+					subtype,
+				} ),
+			} )
+		);
+	}
 
-	const formats = apiFetch( {
-		path: addQueryArgs( '/wp/v2/search', {
-			search,
-			per_page: perPage,
-			type: 'post-format',
-		} ),
-	} );
+	if ( ! type || type === 'term' ) {
+		queries.push(
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/search', {
+					search,
+					per_page: perPage,
+					type: 'term',
+					subtype,
+				} ),
+			} )
+		);
+	}
+
+	if ( ! type || type === 'post-format' ) {
+		queries.push(
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/search', {
+					search,
+					per_page: perPage,
+					type: 'post-format',
+					subtype,
+				} ),
+			} )
+		);
+	}
 
-	return Promise.all( [ posts, terms, formats ] ).then( ( results ) => {
+	return Promise.all( queries ).then( ( results ) => {
```

Comment:
Not introduced by this PR, but this does result in all of the results being smooshed together in the suggestions in a possibly unintended way.

Posts are always first, followed by terms and then post formats, even if the latter is more relevant to the search query.

Maybe this needs to be a REST API change for it work properly. Alternatively, if the Link block only handled external links it wouldn't be as much of an issue.

Reply:
Hm, tough one.

A related problem is that `per_page` doesn't really work because if we set e.g. `per_page = 3` we'll end up with 9 results because we're performing 3 requests.

I'm not really sure how to fix this. Right now the REST API endpoint is architected in a way such that `type` is required and can only be set to one value. Perhaps that needs to change? cc. @TimothyBJacobs 

> Posts are always first, followed by terms and then post formats, even if the latter is more relevant to the search query.

Note that `/wp/v2/search` doesn't order its results by relevancy. But, yes, I get what you mean.

Reply:
We could change the search endpoint to accept multiple types, but that doesn't really solve the problem as far as I can see. Each search handler is distinct, so we can't do one sql query to combine them all. So we'll wind up with the same pagination issue. We can divide the `per_page` by the number of search handlers, but if a search handler doesn't fill its entire "slot" then we'll return an inadequate number of responses.

Reply:
In the upstream branch I made it so that it'll fetch `perPage * 3` results and then slice the array to have only `perPage` entries. This means posts take priority over terms and post formats, which I think is right.